Перевод окна статистики раунда

Ohar opened the pull request:
## Changelog
:cl:
tweak: Перевод окна статистики раунда
/:cl:

![изображение](https://user-images.githubusercontent.com/3854741/159130567-5dcbb006-2219-43a4-be1d-d20139788b49.png)

![изображение](https://user-images.githubusercontent.com/3854741/159130169-40c968ae-f9ba-47fb-94d0-9e053e377995.png)

Ohar committed:
еревод фраз торговых автоматов

Ohar committed:
Фразы торговых автоматов: мелкие правки

Ohar committed:
Merge branch 'ss220-space:master220' into master220

Ohar committed:
Merge branch 'ss220-space:master220' into master220

Ohar committed:
Локализован шведский акцент

Ohar committed:
replacetextEx → replacetextEx_char
See https://github.com/ss220-space/Paradise/pull/647#pullrequestreview-878443663

Ohar committed:
swedish desc: change text

Ohar committed:
Round Statistics and Score translation

Ohar committed:
Merge pull request #1 from ss220-space/master220

Update from SS220

botPurple reviewed (approved):
Проблемы со склонениями при подсчете очков. Но мне обещали, что их исправят когда-то в конце перевода всего парадайза 🤔

Ohar committed:
Fix mistype

Ohar committed:
Merge branch 'master220' into Swedish-l10n

Ohar committed:
Merge branch 'master220' into Swedish-l10n

Ohar committed:
Мелкие правки шведского акцента

Ohar committed:
Merge remote-tracking branch 'origin/Swedish-l10n' into Swedish-l10n

Ohar committed:
УБраны большая часть «ь» и «й», они сильно забивают речь

Ohar committed:
Упрощение шведского акцента

Ohar committed:
Добавлена функция склонения числительных

Ohar committed:
scoreboard: исправлены склонения числительных

Ohar commented:
@botPurple 
Склонения при подсчёте очков исправлены

Ohar committed:
Склонение числительных: Теперь правильно обрабатываются числа больше сотни

Ohar committed:
Склонение числительных: Добавлено значение по-умолчанию

Vallat commented on code/__HELPERS/pronouns.dm:
```diff
@@ -42,6 +42,18 @@
 	if(.)
 		. = "e[.]"
 
+/datum/proc/declension_ru(num, single_name, double_name, multiple_name)
```
Очко очка, а функция не датумная
```suggestion
/proc/declension_ru(num, single_name, double_name, multiple_name)
```

Vallat commented on code/__HELPERS/pronouns.dm:
```diff
@@ -42,6 +42,18 @@
+			if(2,3,4)
+				return double_name
+			else
+				return multiple_name
```
Чуть поменял саму функцию, так должно быть более читаемо. Плюс ко всему, оно теперь будет работать и с нецелыми числами
```suggestion
	if((num in 11 to 20) || (num > 9 && (num % 100) < 20))
		return multiple_name
	var/remainder = num % 10
	if(remainder == 1)
		return single_name
	else if(remainder && remainder < 5)
		return double_name
	return multiple_name
```

Ohar replied:
@Vallat [Поправил](https://github.com/ss220-space/Paradise/pull/650/commits/93d5d960d53fdb6741b569274d49963b67d56aa4)

Vallat replied:
Плохо поправил, не сильно вникая я уже вижу что код нормально не покрывает как минимум интервал (4, 5). Если будет число 4,5 то функция вернёт множественное значение, а должна двойное

Ohar replied:
@Vallat [Исправил обработку четвёрки](https://github.com/ss220-space/Paradise/pull/650/commits/4f2e3e94f9bb56c893fa7d3028b59afa68fb4db6)

Vallat replied:
Не исправил, посмотри на мой код. У тебя отрезок от 2 до 5, а должно быть два интервала: от 0 до 1 и от 1 до 5. Т.е. фактически ты не покрываешь большую долю случаев. Рекомендую просто предложенный код использовать, в котором изначальная логика уже переписан и доработана.

Ohar replied:
Эм, я проверил, [правка](https://github.com/ss220-space/Paradise/pull/650/commits/4f2e3e94f9bb56c893fa7d3028b59afa68fb4db6) была не нужна, код был правильный  😃 
![изображение](https://user-images.githubusercontent.com/3854741/159117453-2b677ecc-9b28-4768-b077-52486d03e04f.png)

Правку [откатил](https://github.com/ss220-space/Paradise/pull/650/commits/44335b4b54c29a84e1808d80aef74a9bde706a98)

Ohar replied:
@Vallat 
Я думаю что склонять надо так, как если бы оно писалось прописью, иначе можно умереть нахуй.

- 0,1 — ноль целых, одна десятая окнА
- 2,2 — две целых, две десятых окнА
- 5,9 — пять целых, девять десятых окнА
- 11,4 — одиннадцать целых, четыре десятых окнА

То есть дробные всегда во множественном числе

[Добавил обработку исключений и проверку целочисленности](https://github.com/ss220-space/Paradise/pull/650/commits/eca4589ab414d0394bf2caf80f7cc7bb19fa9b6e). Дробные числа теперь всегда даю.т множественное число.

Ohar committed:
Update code/__HELPERS/pronouns.dm

Co-authored-by: Vallat <vox.crit2013@gmail.com>

Ohar committed:
Merge branch 'master220' into Round_statistics

Ohar committed:
Merge branch 'master220' into Swedish-l10n

Ohar committed:
Merge remote-tracking branch 'origin/Round_statistics' into Round_statistics

Ohar committed:
Функция склонения оптимизирована и упрощена

Ohar committed:
declension_ru: исправлена обработка чисел, оканчивающихся на «4»

Ohar committed:
Merge branch 'Swedish-l10n' into Round_statistics

Ohar committed:
Правка была не нужна, код был правильный  😃
![изображение](https://user-images.githubusercontent.com/3854741/159117453-2b677ecc-9b28-4768-b077-52486d03e04f.png)

Ohar committed:
declension_ru: обработка исключений, проверка на целое число

Ohar committed:
declension_ru: нецелые числа и не числа → двойственное число

Ohar commented:
@Vallat 
Я думаю что склонять надо так, как если бы оно писалось прописью, иначе можно умереть нахуй.

- 0,1 — ноль целых, одна десятая окнА
- 2,2 — две целых, две десятых окнА
- 5,9 — пять целых, девять десятых окнА
- 11,4 — одиннадцать целых, четыре десятых окнА

То есть дробные **всегда в двойственном числе**

Добавил обработку исключений и проверку целочисленности.
Дробные числа теперь всегда дают двойственное число.

Vallat commented on code/__HELPERS/pronouns.dm:
```diff
@@ -42,6 +42,20 @@
 	if(.)
 		. = "e[.]"
 
+/proc/declension_ru(num, single_name, double_name, multiple_name)
+	try
+		if((isnum(num)) && (round(num) == num))
+			if(((num % 10) == 1) && ((num % 100) != 11)) // 1, not 11
+				return single_name
+			else if(((num % 10) in 2 to 4) && !((num % 100) in 12 to 14)) // 2, 3, 4, not 12, 13, 14
+				return double_name
+			return multiple_name // 5, 6, 7, 8, 9, 0
+		else
+			return double_name // fractional numbers
+	catch()
+		to_chat(usr, "<span class='danger'>An exception has occured during the execution of /proc/declension_ru function.</span>")
+		return ""
```
```suggestion
/proc/declension_ru(num, single_name, double_name, multiple_name)
	if(!isnum(num) || round(num) != num)
		return double_name // fractional numbers
	if(((num % 10) == 1) && ((num % 100) != 11)) // 1, not 11
		return single_name
	if(((num % 10) in 2 to 4) && !((num % 100) in 12 to 14)) // 2, 3, 4, not 12, 13, 14
		return double_name
	return multiple_name // 5, 6, 7, 8, 9, 0
```

Ohar committed:
Update code/__HELPERS/pronouns.dm

Co-authored-by: Vallat <vox.crit2013@gmail.com>

Ohar committed:
Окно итогов раунда: мелкие оптимизации и правки

Ohar committed:
Merge remote-tracking branch 'origin/Round_statistics' into Round_statistics

Ohar committed:
Добыто руды: теперь тоже склоняется

Bizzonium commented on code/game/gamemodes/scoreboard.dm:
```diff
@@ -158,72 +158,72 @@
-	if(SSshuttle.emergency.mode == SHUTTLE_ENDGAME) dat += "<b>Shuttle Escapees:</b> [GLOB.score_escapees] ([GLOB.score_escapees * 25] Points)<br>"
+	<b><u>Общая статистика</u></b><br>
+	<u>Хорошее</u><br>
+	<b>Добыто руды:</b> [GLOB.score_oremined] ([GLOB.score_oremined * 2] очк[declension_ru(GLOB.score_oremined * 2, "о", "а", "ов")])<br>"}
```
Не лучше ли полностью слова использовать? Легче читать и искать
```suggestion
	<b>Добыто руды:</b> [GLOB.score_oremined] ([GLOB.score_oremined * 2] [declension_ru(GLOB.score_oremined * 2, "очко", "очка", "очков")])<br>"}
```

Ohar committed:
Update code/game/gamemodes/scoreboard.dm

Co-authored-by: Aziz Chynaliev <azizonkg@gmail.com>

Ohar committed:
Слова теперь не разбиты на части

Ohar committed:
Merge remote-tracking branch 'origin/Round_statistics' into Round_statistics
